stage1: limit device node creation/reading/writing with DevicePolicy= and DeviceAllow=

iaguis opened the pull request:
This PR limits device node creation/reading/writing using the systemd service options [DevicePolicy=closed](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#DevicePolicy=auto%7Cclosed%7Cstrict) to allow only stardard pseudo devices and [DeviceAllow=](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#DeviceAllow=) to allow devices from user volumes.

This forbids apps in the pod from creating dangerous device nodes like disk partitions.

lucab commented on tests/inspect/inspect.go:
```diff
@@ -78,6 +79,7 @@ var (
 		SilentSigterm      bool
 		CheckMountNS       bool
 		PrintNoNewPrivs    bool
+		CheckMknod         string
```
This is mostly bikeshedding, but I would remove the "check" prefix for this feature, as it is not really checking anything, just doing a plain mknod.

lucab replied:
But I see this stays consistent with other options, so feel free to ignore.

lucab commented on tests/inspect/inspect.go:
```diff
@@ -149,6 +152,45 @@ func main() {
+			os.Exit(1)
+		}
+
+		if err := syscall.Mknod(nodeName, mode, int(majorMinor)); err != nil {
```
Shouldn't this error be bubbled-up in order to exit with failure?

lucab commented on tests/rkt_devices_test.go:
```diff
@@ -0,0 +1,91 @@
+			rktAppArgs:     "",
+			execArgs:       "--check-mknod=c:5:2:/ptmx",
+			expectedOutput: "mknod /ptmx: fail",
+			expectErr:      false,
```
You may want to get an error code from inspect too.

lucab commented on pkg/fileutil/fileutil.go:
```diff
@@ -238,3 +238,13 @@ func IsExecutable(path string) bool {
 }
+
+// IsDeviceNode checks if the given path points to a block or char device.
+func IsDeviceNode(path string) bool {
```
This should mention somewhere (docstring?) that it doesn't follow symlinks on purpose.

lucab commented on stage1/init/common/pod.go:
```diff
@@ -366,6 +366,60 @@ func findBinPath(p *stage1commontypes.Pod, appName types.ACName, app types.App,
+				link := filepath.Join(rktVolumeLinksPath, v.Name.String())
+
+				if _, err := os.Lstat(link); os.IsNotExist(err) {
+					if err := os.Symlink(tgt, link); err != nil {
```
IIRC symlink() doesn't overwrite existing linkpaths. Any reason why you are doing lstat+symlink? In this context it may be fine, but in general it could open for races.

iaguis replied:
Right. There's no reason, I can just do symlink ignore EEXIST

lucab commented on stage1/prepare-app/prepare-app.c:
```diff
@@ -230,6 +230,46 @@ static void mount_sys(const char *root)
+	pexit_if(mkdir(dev_rkt_path, 0700) == -1 && errno != EEXIST,
+		"Failed to create directory \"%s\"", dev_rkt_path);
+
+	volumes_dir = opendir(rkt_volume_links_path);
```
Missing NULL check on return value.

lucab commented on stage1/prepare-app/prepare-app.c:
```diff
@@ -396,6 +436,11 @@ int main(int argc, char *argv[])
+	/* Move symlinks to device node volumes to "/dev/.rkt" so they can be
+	 * used in the DeviceAllow= option of the app's unit file (systemd
+	 * needs the path to start with "/dev". */
+	move_volume_symlinks();
```
This seems slightly misleading to me, as it is not really moving anything around, just adding another layer of symlinks? Has the implementation been changed in the meanwhile?

iaguis replied:
I can rename it to `copy_volume_symlinks()`

lucab commented:
I've left some remarks inline. General comments:
 * Symlinking to /dev/.rkt is really hacky, but I remember the pain with systemd filtering and I don't have better ideas for the moment.
 * How do we plan to let users tweak this? Maybe for the moment we should plug this one too under "paths" insecure-options and then discuss if we need to expose this per-app to users?

tmrts commented on tests/inspect/inspect.go:
```diff
@@ -149,6 +152,45 @@ func main() {
+	if globalFlags.CheckMknod != "" {
+		/* format: c:5:2:name */
+		dev := strings.SplitN(globalFlags.CheckMknod, ":", 4)
+		if len(dev) != 4 {
```
`len(dev) < 4` and `Not enough parameters for mknod` would make the code clearer

iaguis commented:
Updated. PTAL.

- Fixed code review issues.
- Disable device limiting with `--insecure-options=paths`.
- Add some test for `--insecure-options=paths`.

lucab commented on stage1/prepare-app/prepare-app.c:
```diff
@@ -230,6 +230,47 @@ static void mount_sys(const char *root)
+	pexit_if(mkdir(dev_rkt_path, 0700) == -1 && errno != EEXIST,
+		"Failed to create directory \"%s\"", dev_rkt_path);
+
+	pexit_if((volumes_dir = opendir(rkt_volume_links_path)) == NULL && errno != ENOENT,
```
Is this supposed to continue on `~ENOENT`? Why?

iaguis replied:
we don't create the `/rkt/volumes` directory if the user specifies `--insecure-options=paths`

iaguis replied:
I could return if `errno == ENOENT` instead of not failing and relying on `volumes_dir` to be `NULL` but I think this is fine.

lucab replied:
My bad, I didn't see we reach here unconditionally. All fine.

lucab commented:
Fine for me, except for one small question inline. Pinging @tmrts in case he wants to make another pass.

tmrts commented:
@iaguis LGTM after squashing the commits

iaguis commented:
Commits squashed.

alban commented on pkg/fileutil/fileutil.go:
```diff
@@ -238,3 +238,14 @@ func IsExecutable(path string) bool {
+	d, err := os.Lstat(path)
+	if err == nil {
+		m := d.Mode()
+		return m&os.ModeDevice == os.ModeDevice || m&os.ModeCharDevice == os.ModeCharDevice
```
Testing `os.ModeDevice` is enough because in Golang, ModeDevice and ModeCharDevice are not exclusive, as opposed to syscall.S_IFCHR and syscall.S_IFBLK...

See:
https://golang.org/pkg/os/#FileMode
```
ModeDevice             // D: device file
ModeCharDevice         // c: Unix character device, when ModeDevice is set
```
And:
https://golang.org/src/os/stat_linux.go#L18

iaguis replied:
PRs welcome :P

alban commented on stage1/init/common/pod.go:
```diff
@@ -372,6 +372,63 @@ func findBinPath(p *stage1commontypes.Pod, appName types.ACName, app types.App,
+				err := os.Symlink(tgt, link)
+				switch {
+				case err == nil:
+					toShift = append(toShift, link)
```
It seems to add the wrong thing into the array:

```
$ sudo rkt run --debug  --insecure-options=image,ondisk --interactive --no-overlay --private-users --volume=disk,kind=host,source=/dev/loop-control docker://busybox --mount=volume=disk,target=/opt/disk
...
stage1: failed to configure systemd
  └─failed to transform app "busybox" into systemd service
    └─chown /var/lib/rkt/pods/run/3fbecc11-8129-413d-98a3-c55eec309e1f/stage1/rootfs/rkt/volumes/disk: no such file or directory

$ sudo ls -l /var/lib/rkt/pods/run/3fbecc11-8129-413d-98a3-c55eec309e1f/stage1/rootfs/rkt/volumes/disk
lrwxrwxrwx. 1 root root 35 Aug 11 14:30 /var/lib/rkt/pods/run/3fbecc11-8129-413d-98a3-c55eec309e1f/stage1/rootfs/rkt/volumes/disk -> /opt/stage2/busybox/rootfs/opt/disk

```

iaguis replied:
Right, I could swear I tested this, maybe not with a bind-mount

alban commented on stage1/init/common/pod.go:
```diff
@@ -372,6 +372,63 @@ func findBinPath(p *stage1commontypes.Pod, appName types.ACName, app types.App,
+
+	for _, m := range mounts {
+		v := vols[m.Volume]
+		if v.Kind == "host" {
```
To have less indentation, I prefer
```
if v.Kind != "host" {
    continue
}
```

alban commented on pkg/fileutil/fileutil.go:
```diff
@@ -238,3 +238,14 @@ func IsExecutable(path string) bool {
+// IsDeviceNode checks if the given path points to a block or char device.
+// It doesn't follow symlinks.
+func IsDeviceNode(path string) bool {
+	d, err := os.Lstat(path)
```
Is it correct not to follow symlinks?
```
$ ls -l /dev/rtc*
lrwxrwxrwx. 1 root root      4 Jul  8 17:02 /dev/rtc -> rtc0
crw-------. 1 root root 251, 0 Jul  8 17:02 /dev/rtc0
```
With `/dev/rtc0`, this works:
```
$ sudo rkt run --debug  --insecure-options=image,ondisk --interactive --volume=disk,kind=host,source=/dev/rtc0 docker://busybox --mount=volume=disk,target=/opt/the-dev --exec /bin/cat -- /opt/the-dev
```
With `/dev/rtc`, this does not work:
```
$ sudo rkt run --debug  --insecure-options=image,ondisk --interactive --volume=disk,kind=host,source=/dev/rtc docker://busybox --mount=volume=disk,target=/opt/the-dev --exec /bin/cat -- /opt/the-dev
...
cat: can't open '/opt/the-dev': Operation not permitted
```

The mounting logic of nspawn follows links, so I think the `DeviceAllow=` logic should follow links in the same way.

iaguis replied:
I was following [what systemd does](https://github.com/systemd/systemd/blob/v231/src/nspawn/nspawn-register.c#L138) with [is_device_node()](https://github.com/systemd/systemd/blob/v231/src/basic/stat-util.c#L63).

alban commented:
I filed https://github.com/coreos/rkt/issues/3056 to capture the few comments I added.
